doc/user/content/ops/deployment.md

```diff
@@ -12,61 +12,137 @@ issue](https://github.com/MaterializeInc/materialize/issues/new?labels=C-feature
 
 ## Memory
 
-Materialize stores the majority of its state in-memory, and works best when the streamed data
-can be reduced in some way. For example, if you know that only a subset of your rows and columns
-are relevant for your queries, it helps to avoid materializing sources or views until you've
-expressed this to the system (we can avoid stashing that data, which can in some cases dramatically
-reduce the memory footprint).
+Materialize stores the majority of its state in memory, and works best when the
+streamed data can be reduced in some way. For example, if you know that only a
+subset of your rows and columns are relevant for your queries, it helps to avoid
+materializing sources or views until you've expressed this to the system.
+Materialize can then avoid stashing the full set of rows and columns, which can
+in some cases dramatically reduce Materialize's memory footprint.
 
-To minimize the chances that Materialize runs out of memory in a production environment,
-we recommend you make additional memory available to Materialize via a SSD-backed
-swap file or swap partition.
+### Compaction
 
-This is particularly important in Linux and in Docker, where swap may not be automatically
-setup for you:
+To prevent memory from growing without bound, Materialize periodically
+"compacts" data in [arrangements](/overview/api-components#indexes). For
+example, if you have a source that tracks product inventory, you might receive
+periodic inventory updates throughout the day:
+
+```
+(T-shirts, 9:07am, +500)
+(T-shirts, 11:32am, -1)
+(T-shirts, 3:14pm, -2)
+```
+
+Logical compaction will fold historical updates that fall outside the compaction
+window into the state at the start of the window.
+
+```
+(T-shirts, 3:14pm, +497)
+```
+
+Materialize will only perform this compaction on data that falls outside the
+logical compaction window. The default compaction window is 60 seconds, but the
+window can be adjusted via the
+[`--logical-compaction-window`](/cli/#compaction-window) option.
+
+Adjusting the compaction window involves making a tradeoff between historical
+detail and resource usage. A larger compaction window retains more historical
+detail, but requires more memory. A smaller compaction window uses less memory
+but also retains less historical detail. Larger compaction windows also increase
+CPU usage, as more detailed histories require more compute time to maintain.
+
+{{< warning >}}
+Setting the compaction window too small can prevent Materialize from finding a
+suitable timestamp at which to execute a `SELECT` query, especially if that
+query joins multiple sources together.
+{{< /warning >}}
+
+Note that compaction is triggered in response to updates arriving. As a result,
+if updates stop arriving for a source, Materialize may never compact the source
+fully. This can result in higher-than-expected memory usage.
```

Comment:
At the risk of adding too much detail, this is where `DIFFERENTIAL_EAGER_MERGE` shows up. The text is not wrong, but adding that env var greatly reduces the circumstances in which compaction doesn't occur (to those where complete compaction occurred within the logical compaction window).

Mostly trying to avoid introducing the idea that one should continually change their data or risk memory growth.

Reply:
As above, definitely down to teach users about eager merges, but going to hold off on making any specific recommendation as to the value of `DIFFERENTIAL_EAGER_MERGE` in this PR. I'm not familiar with what a "good" value for this variable is. I also don't love publicizing a differential-branded environment variable (#1223). It's one thing for it to be a secret tuning knob that folks in the know can toggle, but a bit weird to publicly document. I'd advocate for adding a `--eager-compaction` CLI flag and `MATERIALIZE_EAGER_COMPACTION` env var that we then document.

Also, if it's great to have it turned on, should we have it turned on by default?